### DISCUSSION TITLE
Make Web Search Use Multiple Queries

### DISCUSSION THREAD

**Discussion by reecelikesramen** [open-webui/open-webui#7948] [Ideas]
# Feature Request

**Is your feature request related to a problem? Please describe.**
The web search feature only searches a single query, but the prompt (even the default one) generates multiple queries.

**Describe the solution you'd like**
I would like the web search feature to use all the search queries generated.

**Describe alternatives you've considered**
- NUM_SEARCHES * NUM_QUERIES? or (NUM_SEARCHES / NUM_QUERIES) per query and allocate remainder to query[0]?
- default web query prompt could be told to just make 1 good relevant query.

**Additional context**
Right now the web search feature takes in a list of queries, and it's hard coded to select the first in the list. This could be leading to poorer search quality, as the default prompt is encouraged to make several queries, and not necessarily put the best one first. Here's the relevant code:
[open-webui](https://github.com/open-webui/open-webui/tree/29a271959556743e6deb4d55a5a982983335d7ab)/[src](https://github.com/open-webui/open-webui/tree/29a271959556743e6deb4d55a5a982983335d7ab/src)/[lib](https://github.com/open-webui/open-webui/tree/29a271959556743e6deb4d55a5a982983335d7ab/src/lib)/[components](https://github.com/open-webui/open-webui/tree/29a271959556743e6deb4d55a5a982983335d7ab/src/lib/components)/[chat](https://github.com/open-webui/open-webui/tree/29a271959556743e6deb4d55a5a982983335d7ab/src/lib/components/chat)/Chat.svelte
```py
...
                 const prompt = userMessage.content;
		let queries = await generateQueries(
			localStorage.token,
			model,
			messages.filter((message) => message?.content?.trim()),
			prompt
		).catch((error) => {
			console.log(error);
			return [prompt];
		});

		if (queries.length === 0) {
			responseMessage.statusHistory.push({
				done: true,
				error: true,
				action: 'web_search',
				description: $i18n.t('No search query generated')
			});
			history.messages[responseMessageId] = responseMessage;
			return;
		}

                 // RELEVANT LINE HERE
		const searchQuery = queries[0];
...
```

**Comment by aaltulea** [comment]
i was trying to achieve this. the code below will search 3 unique queries. you can change the number of queries by changing 3 in `.slice(0, 3)`. overall, i didn't find it very useful.


```

	const getWebSearchResults = async (
		model: string,
		parentId: string,
		responseMessageId: string
	) => {
		// TODO: move this to the backend
		const responseMessage = history.messages[responseMessageId];
		const userMessage = history.messages[parentId];
		const messages = createMessagesList(history.currentId);

		responseMessage.statusHistory = [
			{
				done: false,
				action: 'web_search',
				description: $i18n.t('Generating search queries')
			}
		];
		history.messages[responseMessageId] = responseMessage;

		const prompt = userMessage.content;
		let queries = await generateQueries(
			localStorage.token,
			model,
			messages.filter((message) => message?.content?.trim()),
			prompt
		).catch((error) => {
			console.log(error);
			return [prompt];
		});

		if (queries.length === 0) {
			responseMessage.statusHistory.push({
				done: true,
				error: true,
				action: 'web_search',
				description: $i18n.t('No search query generated')
			});
			history.messages[responseMessageId] = responseMessage;
			return;
		}

		// Take the first three unique queries
		const uniqueQueries = [...new Set(queries)].slice(0, 3);
		const combinedResultsSet = new Set(); // Ensure unique combined results
		let searchedCount = 0;

		for (let i = 0; i < uniqueQueries.length; i++) {
			const searchQuery = uniqueQueries[i];

			responseMessage.statusHistory.push({
				done: false,
				action: 'web_search',
				description: $i18n.t(
					`Searching: "{searchQuery}" (${i + 1}/${uniqueQueries.length})`
				)
			});
			history.messages[responseMessageId] = responseMessage;

			const results = await processWebSearch(localStorage.token, searchQuery).catch((error) => {
				console.log(error);
				toast.error(error);
				return null;
			});

			if (results) {
				results.filenames.forEach((filename) => combinedResultsSet.add(filename)); // Add to Set
				searchedCount += results.filenames.length;

				if (!responseMessage.files) {
					responseMessage.files = [];
				}

				responseMessage.files.push({
					collection_name: results.collection_name,
					name: searchQuery,
					type: 'web_search_results',
					urls: results.filenames
				});
			}
		}

		const combinedResults = [...combinedResultsSet]; // Convert Set to Array

		if (combinedResults.length > 0) {
			responseMessage.statusHistory.push({
				done: true,
				action: 'web_search',
				description: $i18n.t('Searched {{count}} sites', { count: searchedCount }),
				urls: combinedResults
			});
		} else {
			responseMessage.statusHistory.push({
				done: true,
				error: true,
				action: 'web_search',
				description: 'No search results found'
			});
		}

		history.messages[responseMessageId] = responseMessage;
	};
```

**Comment by reecelikesramen** [comment]
May I ask what search engine and query generator model you're using?

I noticed because I was trying to improve the web search prompt. I realized that if I asked a multipart question it would split up the queries like [q1, q2, q3] if I had 3 different topics, but it would only search q1, and the model would either not answer q2 or q3, or use no context to answer those questions.

I think it's worth having this issue here at least to come to a conclusion on what should be default: just 1 query, or multiple.

**Comment by aaltulea** [comment]
I am using qwen2.5 (7b and 32b), with searxng. In the settings, i specified 20 search results to read. Using the code I posed, the model was able to read 40-60 pages, but at the end it will only pick around 5 to render the answer, which is the same as if it ran one query and read 20 pages. I think if we were able to refine the prompt for search query generation, we will get better results.

**Comment by hurxxxx** [comment]
Has there been any decision made regarding this idea?

Currently, web search still uses only a single hardcoded query even when multiple search queries are generated. I have implemented the following changes to address this:

Added an "Allow Multiple Search Queries" option under Admin Settings → Interface.
If set to False, the behavior remains the same: only the first query generated by the Query Generation Prompt is searched and added to the context.
If set to True, all the generated search queries are executed and added to the context.

![AllowMultipleSearchQueries](https://github.com/user-attachments/assets/596f3c70-2294-4103-9527-a2ddca575fbf)


The changes are as follows:

- Removed duplicate keywords
- Added code related to the "Allow Multiple Search Queries" option
- Added a for loop

```python

async def chat_web_search_handler(
    request: Request, form_data: dict, extra_params: dict, user
):

.......... 

queries = queries.get("queries", [])
queries = list(dict.fromkeys(queries))   #Remove duplicate keywords


#searchQuery = queries[0]  ##Disable

# 
 searchQueries = queries
 if not request.app.state.config.ALLOW_MULTIPLE_SEARCH_QUERIES:
     searchQueries = [queries[0]]

 for searchQuery in searchQueries:

    ....................

``` 

Please share your thoughts. I believe this feature is absolutely necessary.

**Reply by Classic298** [reply]
This is an amazing feature IMO! Perhaps you can even add a simple input box which allows how many queries can be executed at maximum (e.g. 3 as the default value)

Make a PR! :D

**Reply by reecelikesramen** [reply]
I think multiple search queries are a must have. I don't get very good results with the search feature as-is.

**Reply by hurxxxx** [reply]
Please refer to PR #8816 for more details.
